src/jwe.c

```diff
@@ -3340,17 +3358,32 @@ int r_jwe_decrypt_payload(jwe_t * jwe) {
   size_t payload_enc_len = 0, unzip_len = 0;
   unsigned char tag[128];
   size_t tag_len = 0;
+  size_t ciphertext_b64_len;
   int cipher_cbc;
   struct _o_datum dat = {0, NULL}, dat_ciph = {0, NULL}, dat_tag = {0, NULL};
 
-  if (jwe != NULL && jwe->enc != R_JWA_ENC_UNKNOWN && !o_strnullempty((const char *)jwe->ciphertext_b64url) && !o_strnullempty((const char *)jwe->iv_b64url) && jwe->key != NULL && jwe->key_len && jwe->key_len == _r_get_key_size(jwe->enc)) {
+  if (jwe != NULL && jwe->enc != R_JWA_ENC_UNKNOWN && (ciphertext_b64_len = o_strlen((const char *)jwe->ciphertext_b64url)) != 0 && !o_strnullempty((const char *)jwe->iv_b64url) && jwe->key != NULL && jwe->key_len && jwe->key_len == _r_get_key_size(jwe->enc)) {
+    /* ensure payload_enc_buflen is a multiple of cipher_block_size
+     * if the cipher is a block-mode cipher
+     */
+    if (gnutls_is_block_cipher(_r_get_alg_from_enc(jwe->enc))) {
+      size_t ciphertext_decoded_len = (ciphertext_b64_len * 3) / 4;
```

Comment:
I prefer using `o_base64url_decode` to calculate the output length of a base64url transformation:

```C
ciphertext_decoded_len = o_base64url_decode(jwe->ciphertext_b64url, ciphertext_b64_len, NULL, &ciphertext_b64_len);
```

Reply:
Hi Nicolas,

Thank you for your careful review. I have implemented the changes you requested.

This particular change, using o_base64url_decode to calculate the output length seemed unnecessary duplication of decoding a string, since `o_base64_decode_agnostic()`  completely decodes the string, also checking the validity of each character. This is one reason why my fork will diverge from your work, as I make performance-related optimizations in my fork.

The divergence will make it more difficult to share future improvements, because attempting to merge improvements to diverging implementation areas will encounter merge conflicts, and thus would require porting from one fork to the other.

It is very important for the project where I currently use your rhonabwy library to be able to verify and decrypt JWEs while under DDOS load, since distinguishing legitimate from attack traffic is the critical path. I imagine the performance of `r_jwe_decrypt_payload` would be important to any project that processes authentication JWEs.

Reply:
I understand your concern, but the formula to calculate the decoded value may be incorrect then, because in C, integer division truncates the result and skips the decimal.

Reply:
The integer truncation has the desired, and correct effect. There are 4 distinct cases, where the base64 encoded length modulo 4 is one of 0, 1, 2 or 3:
<pre>
bm=b64_len % 4    (bm*3/4)
(num input bytes) (num output bytes)
0                 0
1                 0
2                 1
3                 2
</pre>

3 of these cases are of well formed input (b64_len % 4 is either 0, 2 or 3), and result in a decoded length of 0, 1 or 2 bytes respectively, and 1 case is malformed input (b64_len % 4 is 1), where 1 byte of input is insufficient to form a full byte of output, which means the b64 string is corrupted. In the malformed input case, the 1 byte of input is correctly calculated to result in 0 bytes of output.

Reply:
Arguably, the calculation `(ciphertext_b64_len * 3) / 4` does not check if the base64 input is well formed (ie, length mod 4 != 1), which `o_base64url_decode()` does. So in the implementation above, a malformed input will still be decoded (attempted), though not decrypted. This attempt could be avoided by also checking for well-formed input.

For my own fork, I will implement a check similar to `check_jwe_b64_input_valid_and_wellformed()` from https://godbolt.org/z/vdKodfWYs, which is very fast on any CPU, avoids the div instruction, which is slow on some CPUs (doesn't affect my application, but you might care 😄), and is quite readable. It checks that the input is non-zero length, not malformed from a base64 pov and that it is a multiple of block size, in only a few assembly instructions.